libc: add support for custom streams with fopencookie()

michallenc opened the pull request:
## Summary
This commit adds support for custom stream via `fopencookie` function. The function allows the programmer the create his own custom stream for IO operations and hook his custom functions to it.

This is a non POSIX interface defined in Standard C library and implemented according to it. The only difference is in usage of `off_t` instead of `off64_t`. Programmer can use 64 bits offset if `CONFIG_FS_LARGEFILE` is enabled. In that case `off_t` is defined as `int64_t` (`int32_t` otherwise).

The interface will be useful for future fmemopen implementation.

## Impact
New interface.

## Testing
Tested with [example application on man page](https://man7.org/linux/man-pages/man3/fopencookie.3.html) on SAMv7 MCU.

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,229 @@
+
+struct cookie_s
+{
+  void *cookie;                    /* pointer to the caller's cookie struct */
```
add FAR

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,229 @@
+struct cookie_s
+{
+  void *cookie;                    /* pointer to the caller's cookie struct */
+  cookie_io_functions_t cookie_io; /* programmer-defined hook functions */
```
ALL first char in comment change to upper case

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -425,6 +425,27 @@ struct inode
+typedef ssize_t cookie_read_function_t(void *cookie, char *buf, size_t size);
+typedef ssize_t cookie_write_function_t(void *cookie, const char *buf,
+                                        size_t size);
+typedef ssize_t cookie_seek_function_t(void *cookie, off_t *offset,
```
```suggestion
typedef int cookie_seek_function_t(void *cookie, off_t *offset,
```

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -425,6 +425,27 @@ struct inode
+                                        size_t size);
+typedef ssize_t cookie_seek_function_t(void *cookie, off_t *offset,
+                                       int whence);
+typedef ssize_t cookie_close_function_t(void *cookie);
```
```suggestion
typedef int cookie_close_function_t(void *cookie);
```

xiaoxiang781216 commented on include/stdio.h:
```diff
@@ -244,6 +244,11 @@ int    dprintf(int fd, FAR const IPTR char *fmt, ...) printf_like(2, 3);
 
+/* Custom stream operation fopencookie. */
+
+FAR FILE *fopencookie(void *cookie, const char *mode,
```
```suggestion
FAR FILE *fopencookie(FAR void *cookie, FAR const char *mode,
```

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -425,6 +425,27 @@ struct inode
+
+typedef struct
+{
+  cookie_read_function_t *read;
```
add FAR for ALL pointers

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,229 @@
+static ssize_t freadcookie(FAR struct file *filep, FAR char *buf,
+                           size_t nbytes)
+{
+  struct cookie_s *cookie = (struct cookie_s *)filep->f_priv;
```
```suggestion
  struct cookie_s *cookie = (FAR struct cookie_s *)filep->f_priv;
```
add FAR to ALL pointers

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,229 @@
+
+  /* And get file descriptor. */
+
+  fd = file_allocate_from_tcb(nxsched_self(), filep.f_inode, filep.f_oflags,
```
libc can't call os internal function directly

michallenc replied:
If I create `fs/vfs/fs_fopencookie.c` with function `fs_fopencookie` that would call the internal function (similarly to `fs_fdopen`), would that be ok?

xiaoxiang781216 replied:
I think we don't need allocate fd here, it's enough to set fs_fd to -1, just like how musl implement fopencookie:
https://github.com/bminor/musl/blob/master/src/stdio/fopencookie.c#L121

michallenc replied:
But I need valid fd to be able to call proper functions, no? `fopencookie` returns `FILE` structure which contains file descriptor. Then if IO operation (lets say `write`) is called, NuttX uses the file descriptor to get `struct file` which contains pointer to `inode` which has `u.i_ops` with IO operations. And this operations lead to private functions in `lib_fopencookie.c`. 

MUSL implementation is a bit different, they have pointers to IO directly in `FILE`, we don´t. So I think we always need file descriptor if we want to get there from `FILE` stream.

xiaoxiang781216 replied:
Ok, you implementation reuse file_operation_s infrastructure, but it mayn't work in protected/kernel mode, since fclosecookie... is part of kernel, but call the callback provided by userspace directly.

gregory-nutt replied:
> Ok, you implementation reuse file_operation_s infrastructure, but it mayn't work in protected/kernel mode, since fclosecookie... is part of kernel, but call the callback provided by userspace directly. I would suggest that you have to add these callback to file_struct and call them from libc/stdio directly.
> 
> @patacongo what are you thinking?

_[I have not looked at this particular PR, but I can give my usual response when callbacks from the OS are suggested]_

Direct callbacks from the OS into user code have always been forbidden.  I don't think we should permit that now.  Some justifications:

- This is not the POSIX way to do things.  There is no way to do such things under any POSIX OS.  We have always been concerned about POSIX compliant and portability.  Linux code is often ported to NuttX. NuttX code should be runnable on any any POSIX OS as well (with the appropriate drivers).  To accomplish these goals, callbacks must not be used.

- The callback would probably be running on an kernel stack and not the user stack.  In that case, TLS data on the stack would not be accessible.  The callback were asynchronous with the tasking.  A callback is not a context switch; the interrupt user task is still in place and the OS thinks that is that task that is running.  So any OS interactions in this bastardized state could likely be fatal.

- The callback would run at whatever the priority of the calling logic is running at.  This will break real time behavior which must be strict priority-based scheduling.

There are several differ build modes for NuttX:  The FLAT build, the PROTECTED build for better security, and the KERNEL build for full support of processes much like Linux.  In the PROTECTED and KERNEL builds, there are more significant technical issues.

There is some documentation of at least the PROTECTED build here:  https://cwiki.apache.org/confluence/display/NUTTX/NuttX+Protected+Build

- In the PROTECTED and KERNEL build modes, the kernel runs in supervisor mode in a supervisor address space while all user applications run in non-privileged user mode in a user address space.  In the PROTECTED mode, you _could_ call from the kernel supervisor code into user code but this would, however, be a gaping security hole and, hence, must never be done.  Potentially hostile user-supplied code must never run in supervisor mode in this build.

- The KERNEL build differs in that each _process_ has its own address environment and, in this build, the system will crash if such a thing is attempted.  That is because the notification arrives asynchronously with tasking; you never know which process is running and you would call into the wrong address space.  This could be fixed be changing the address environment prior to each callback, that would have big performance and complexity issues.  It is better to do things in the POSIX way.

So even if you use only the FLAT build, direct callbacks from the OS into the user code must be prohibited.  What is the POSIX way of doing an aynchronous notification?  Signals.  In a POSIX OS, you should use standard POSIX signaling and avoid even thinking about callbacks.

Drivers under drivers/ that do asynchronous notifications use signals for this purpose.  Consider drivers/input/button_upper.c.  This is a simple button driver that sends a notification to a user task when a button is pressed or release:

- Look at the function btn_ioctl().  The BTNIOC_REGISTER command is used to register a notification.  The process ID of the user thread is provided and retained for the notification.

- When the button press event occurs, the notification signal is sent in btn_sample().

- A signal handler in the user thread that registered for the notification can then run, in the correct task context, in user mode, and at the correct priority assigned to the task.

xiaoxiang781216 replied:
Since stdio is a pure usespace library, I would suggest keep all fopencookie related stuff in libc to avoid the complexity of kernel/userspace interactions.
@michallenc you can achieve this by:

1. Add cookie_io_functions_t to file_struct(FILE)
2. Assign the internal callback which read/write/seek through fd to cookie_io_functions_t in fopen
3. Assign caller supplied callback in fopencookie
4. Replace read/write/seek with cookie_io_functions_t's one

patacongo replied:
> Since stdio is a pure usespace library, I would suggest keep all fopencookie related stuff in libc to avoid the complexity of kernel/userspace interactions.

That would not solve the technical problem.  It eliminates the kernal/user privilege issue, but is otherwise insufficient.  That is because the caller executes in Task/Process A but the user-implementation of the file is implemented in a different Task/Process B.  There are lots of problems with that:

- You would be executing code in Task B in the context of Task A.
- That means Task B would use Task A's stack and file structures, not those of Task B. 
- The Task B file system code could not access Task B's TLS data .
- From the standpoint of the OS, Task A is running, not Task B.  Task B could not access the OS as Task B and many things just would not work properly.
- The logic in Task B would run with with priority and identity of Task A.
- It would not work at all in the KERNEL build mode because Task B resides in a different process address space than does Task A.  That would crash immediately and dramatically.
- etc. Many more issues.

I don't believe that any of that behavior is acceptable.

I still believe that the only viable solution must depend on using a proper POSIX IPC like signals, message queues, fifos, local networks, etc.  Then Task B runs in the proper context and address space of Task B.

xiaoxiang781216 replied:
FILE is private to a specific task/process, it's illegal to share FILE * between the different tasks. So, @patacongo the scenario you describe can never happen.

michallenc replied:
> Add cookie_io_functions_t to file_struct(FILE)
Assign the internal callback which read/write/seek through fd to cookie_io_functions_t in fopen
Assign caller supplied callback in fopencookie
Replace read/write/seek with cookie_io_functions_t's one

I am not sure this would solve the problem. Because the callback to `fopencookie` would still go through `fs/vfs`. Now this would work in FLAT build but probably not in protected build. What we would need is that `libc` would call the callbacks directly for `FILE *` if custom streams are defined as it is done in ucLibc for example https://elixir.bootlin.com/uclibc-ng/v1.0.7/source/libc/stdio/fread.c#L68 This way we would not need  But that is too big change now.

xiaoxiang781216 replied:
> > Add cookie_io_functions_t to file_struct(FILE)
> > Assign the internal callback which read/write/seek through fd to cookie_io_functions_t in fopen
> > Assign caller supplied callback in fopencookie
> > Replace read/write/seek with cookie_io_functions_t's one
> 
> I am not sure this would solve the problem. Because the callback to `fopencookie` would still go through `fs/vfs`.

In my suggestion, you should stop to implement file_operation_s and call file_allocate_from_tcb. How do you go through fs/vfs?

> Now this would work in FLAT build but probably not in protected build. What we would need is that `libc` would call the callbacks directly for `FILE *` if custom streams are defined as it is done in ucLibc for example https://elixir.bootlin.com/uclibc-ng/v1.0.7/source/libc/stdio/fread.c#L68 This way we would not need But that is too big change now.

This is the right way to go. And most change is minor.

michallenc replied:
> In my suggestion, you should stop to implement file_operation_s and call file_allocate_from_tcb. How do you go through fs/vfs without them?

Yes, but you mention 

> Assign the internal callback which read/write/seek through fd to cookie_io_functions_t in fopen

How would I get fd without actually using `file_allocate_from_tcb`? I cannot even call `fopen` without that.

xiaoxiang781216 replied:
fd shouldn't reference in common code, but in the callback provided by stdio self. Actually fd should change to void * instead.

gregory-nutt commented on include/stdio.h:
```diff
@@ -244,6 +244,11 @@ int    dprintf(int fd, FAR const IPTR char *fmt, ...) printf_like(2, 3);
+
+FAR FILE *fopencookie(FAR void *cookie, FAR const char *mode,
+                      cookie_io_functions_t io_funcs);
+
```
It is prohibited to add non-standard interfaces to the OS.  We are purely standards based and non-standard, ad hoc OS interfaces are generally prohibited.

xiaoxiang781216 replied:
it is implemented by *BSD too:
https://man.freebsd.org/cgi/man.cgi?query=fopencookie&apropos=0&sektion=3&manpath=FreeBSD+11-current&format=html

michallenc replied:
Yes, I think the usage in Linux is a justification here although it is not a POSIX interface. Morover this will be helpful in the implementation of `fmemopen` https://pubs.opengroup.org/onlinepubs/9699919799/functions/fmemopen.html (although we can do it without `fopencookie` but it complicates things a bit).

gregory-nutt commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,229 @@
+
+  /* And get file descriptor. */
+
+  fd = file_allocate_from_tcb(nxsched_self(), filep.f_inode, filep.f_oflags,
```
I am strongly opposed to adding any user interfaces to the OS that hand no basis in established standards.  This is a terrible practice and should not be accepted.

gregory-nutt commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,229 @@
+  DEBUGASSERT(cookie->cookie != NULL && cookie->cookie_io.read != NULL);
+
+  ret = cookie->cookie_io.read(cookie->cookie, buf, nbytes);
+  return ret;
```
Are these the user space callbacks you were referring to?  Yes, these are wrong.  This can't work this way there can be no such calls from the kernel address space.

This is similar to the way that the usrsock works.  The usrsock functions are "called" from the OS, but not directly.   Rather, the call data is marshaled through a message interface to the user code.  This is way this should be done as well.  The interface is described in include/nuttx/net/usrsock.h

At one time I wrote some user-space device drivers (character and block) based on the BSD user space drivers, but I don't think those made it into the repository.  The design was similar to the usrsock design.

michallenc replied:
> Are these the user space callbacks you were referring to? Yes, these are wrong. This can't work this way there can be no such calls from the kernel address space.

I think `stdio` library is purely user space (and protect build builds it twice, one for kernel and one for user space) so this callback here should be correct. The issue @xiaoxiang781216 is reffering to is that the route from user call to the callback is: `fseek` from app -> `fseeko` in libc -> `lseek` in `fs/vfs` -> `fseekcookie` in libc. I think all we need to do is to avoid the usage of  `lseek` in `fs/vfs` and this way we ensure we stay in user space.

gregory-nutt commented:
I personally think you are going down the wrong path with this ad hoc logic.  This is very high risk because I don't think either of you are considering all of the issues and complexities.  You apparently seem focused on avoiding a clean context switch (which I believe is the correct, clean, flexible solution).

I believe we need to at least insist on a good test for this code to give people like me confidence.  A good test would include:

- All build modes:  FLAT, PROTECTED, and KERNEL (we would need some testing help).
- Assure that user code never executes in kernel mode in the PROTECTED build
- Test with client and server in same task/process; Client and server in different tasks/processes and different address spaces (in KERNEL mde).

If it could pass a pretty rigorous test then I would become a believer.  That, I think is a requirement for merging.  I am not a believer now.

mu578 commented:
I once look at that to handle a fomempen and alike properly, it would require that struct FILE not to be a direct alias to internal FS struct, as indeed you need to have reference to callbacks and flags, so it would require an indirection where struct FILE holds a private pointer to internal.

michallenc commented:
> You apparently seem focused on avoiding a clean context switch (which I believe is the correct, clean, flexible solution).

@patacongo  My main argument againts hooks you described above (various notifications, events, signal handlers etc.) is that it would lead to additional requirements for user application (different from how `fopencookie` is used in GNU/Linux). I havent had time to go through all code you suggested (graphics etc) so maybe I am missing something. But I think we should keep it in simple `fseak`, `fputs` etc operations.

@mu578 I like the structure you proposed, that would solve a lot of issues with current `file_struct` and as @xiaoxiang781216 mentioned it should not be that difficult (mostly renaming etc.). We could still keep pointers to read/write/seek at `file` in `fs/vfs` as it is now for drivers and just add the possibility for custom streams if required.

I would do this primarily for `fopencookie` and than build `fmemopen` and `open_memstream` on that if the community would agree as I think it is still beneficial to have `fopencookie` implemented (and its presence in Linux should justify that in my opinion).

I also think we should probably move this discussion to mailing list now.

xiaoxiang781216 commented:
> I once had a look at that to handle a `fomempen` and alike properly, it would require that LIBC `struct FILE` not to be a direct alias to internal FS struct, as indeed you need to keep a reference to fs operation callbacks and permission flags, so it would require an indirection where `struct FILE` holds a private pointer to internal. Meanwhile, I think LIBC stdio would benefit at term that indirection to be made.
> 
> ```c
> struct libc_stdio_file
> {
>   // ... stdio specifics;
>   libc_stdio_fread_t  read;
>   libc_stdio_fwrite_t write;
>   libc_stdio_fseek_t  seek;
>   libc_stdio_fclose_t close;
>   libc_stdio_cookie_t cookie;
> 
>   file_struct *       internal;
> }
> 
> typedef libc_stdio_file FILE;
> ```

we don't need add a new struct libc_stdio_file. Here is the current definition:
```
struct file_struct
{
  FAR struct file_struct *fs_next;      /* Pointer to next file stream */
  rmutex_t                fs_lock;      /* Recursive lock */
  int                     fs_fd;        /* File descriptor associated with stream */
#ifndef CONFIG_STDIO_DISABLE_BUFFERING
  FAR unsigned char      *fs_bufstart;  /* Pointer to start of buffer */
  FAR unsigned char      *fs_bufend;    /* Pointer to 1 past end of buffer */
  FAR unsigned char      *fs_bufpos;    /* Current position in buffer */
  FAR unsigned char      *fs_bufread;   /* Pointer to 1 past last buffered read char. */
#  if CONFIG_STDIO_BUFFER_SIZE > 0
  unsigned char           fs_buffer[CONFIG_STDIO_BUFFER_SIZE];
#  endif
#endif
  uint16_t                fs_oflags;    /* Open mode flags */
  uint8_t                 fs_flags;     /* Stream flags */
#if CONFIG_NUNGET_CHARS > 0
  uint8_t                 fs_nungotten; /* The number of characters buffered for ungetc */
  unsigned char           fs_ungotten[CONFIG_NUNGET_CHARS];
#endif
};

typedef struct file_struct FILE;
```
Here is the simple change to the struct:
```
struct file_struct
{
  ...
-  int                     fs_fd;        /* File descriptor associated with stream */
+  cookie_io_functions_t fs_iofunc;
+  FAR void *            fs_cookie;
```
};
```
and replace all read/write/seek/close with stream->iofunc.read/write/seek/close.
The remaining work is adding the simple wrapper for fopen:
```
static ssize_t read_fd(void *cookie, char *buf, size_t size)
{
  int fd = (int)(intptr_t)cookie;
  return read(fd, buf, size);
}

static ssize_t write(void *cookie, const char *buf, size_t size)
{
  int fd = (int)(intptr_t)cookie;
  return write(fd, buf, size);
}

static int seek(void *cookie, off_t *offset, int whence)
{
  int fd = (int)(intptr_t)cookie;
  off_t ret = seek(fd, *offset, whence);
  if (ret >= 0)
    {
      *offset = ret;
      ret = 0;
    }

  return ret;
}
```

gregory-nutt commented:
> > You apparently seem focused on avoiding a clean context switch (which I believe is the correct, clean, flexible solution).
> 
> @patacongo My main argument againts hooks you described above (various notifications, events, signal handlers etc.) is that it would lead to additional requirements for user application (different from how `fopencookie` is used in GNU/Linux). I havent had time to go through all code you suggested (graphics etc) so maybe I am missing something. But I think we should keep it in simple `fseak`, `fputs` etc operations.

The marshaled data should be indisguishable from a direct function (provided an intermediate marshaling/de-marshaling,.

My main argument against this approach is that I have not heard any solution that will actually work in all conditions and build modes.

xiaoxiang781216 commented:
> > > You apparently seem focused on avoiding a clean context switch (which I believe is the correct, clean, flexible solution).
> > 
> > 
> > @patacongo My main argument againts hooks you described above (various notifications, events, signal handlers etc.) is that it would lead to additional requirements for user application (different from how `fopencookie` is used in GNU/Linux). I havent had time to go through all code you suggested (graphics etc) so maybe I am missing something. But I think we should keep it in simple `fseak`, `fputs` etc operations.
> 
> The marshaled data should be indisguishable from a direct function (provided an intermediate marshaling/de-marshaling,.
> 
> My main argument against this approach is that I have not heard any solution that will actually work in all conditions and build modes.

@patacongo I think @michallenc initial design may confuse you. stdio is a pure userspace library, the implementation of fopencookie or fmemopen doesn't need involve any kernel stuff: it's enough to add the callback(read/write/seek/close) to struct file_struct(FILE).
So, I would suggest you read the manual from:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/fmemopen.html
https://man7.org/linux/man-pages/man3/fopencookie.3.html

mu578 commented:
> > I once had a look at that to handle a `fomempen` and alike properly, it would require that LIBC `struct FILE` not to be a direct alias to internal FS struct, as indeed you need to keep a reference to fs operation callbacks and permission flags, so it would require an indirection where `struct FILE` holds a private pointer to internal. Meanwhile, I think LIBC stdio would benefit at term that indirection to be made.
> > ```c
> > struct libc_stdio_file
> > {
> >   // ... stdio specifics;
> >   libc_stdio_fread_t  read;
> >   libc_stdio_fwrite_t write;
> >   libc_stdio_fseek_t  seek;
> >   libc_stdio_fclose_t close;
> >   libc_stdio_cookie_t cookie;
> > 
> >   file_struct *       internal;
> > }
> > 
> > typedef libc_stdio_file FILE;
> > ```
> 
> we don't need add a new struct libc_stdio_file. Here is the current definition:
> 
> ```
> struct file_struct
> {
>   FAR struct file_struct *fs_next;      /* Pointer to next file stream */
>   rmutex_t                fs_lock;      /* Recursive lock */
>   int                     fs_fd;        /* File descriptor associated with stream */
> #ifndef CONFIG_STDIO_DISABLE_BUFFERING
>   FAR unsigned char      *fs_bufstart;  /* Pointer to start of buffer */
>   FAR unsigned char      *fs_bufend;    /* Pointer to 1 past end of buffer */
>   FAR unsigned char      *fs_bufpos;    /* Current position in buffer */
>   FAR unsigned char      *fs_bufread;   /* Pointer to 1 past last buffered read char. */
> #  if CONFIG_STDIO_BUFFER_SIZE > 0
>   unsigned char           fs_buffer[CONFIG_STDIO_BUFFER_SIZE];
> #  endif
> #endif
>   uint16_t                fs_oflags;    /* Open mode flags */
>   uint8_t                 fs_flags;     /* Stream flags */
> #if CONFIG_NUNGET_CHARS > 0
>   uint8_t                 fs_nungotten; /* The number of characters buffered for ungetc */
>   unsigned char           fs_ungotten[CONFIG_NUNGET_CHARS];
> #endif
> };
> 
> typedef struct file_struct FILE;
> ```
> 
> Here is the simple change to the struct:
> 
> ```
> struct file_struct
> {
>   ...
> -  int                     fs_fd;        /* File descriptor associated with stream */
> +  cookie_io_functions_t fs_iofunc;
> +  FAR void *            fs_cookie;
>   ...
> };
> ```
> 
> and replace all read/write/seek/close with stream->iofunc.read/write/seek/close. The remaining work is adding the simple wrapper for fopen:
> 
> ```
> static ssize_t read_fd(void *cookie, char *buf, size_t size)
> {
>   int fd = (int)(intptr_t)cookie;
>   return read(fd, buf, size);
> }
> 
> static ssize_t write_fd(void *cookie, const char *buf, size_t size)
> {
>   int fd = (int)(intptr_t)cookie;
>   return write(fd, buf, size);
> }
> 
> static int seek_fd(void *cookie, off_t *offset, int whence)
> {
>   int fd = (int)(intptr_t)cookie;
>   off_t ret = seek(fd, *offset, whence);
>   if (ret >= 0)
>     {
>       *offset = ret;
>       ret = 0;
>     }
> 
>   return ret;
> }
> ```

you would if want to keep separated internal implementation to actual in-memory Os layout and a libc stdio frontend. Indeed `nutxx/fs.h` is one possible core-stream handling implementation, you don't want to mixup ; theoretically you could swap to any `OS vfs` implementation without changing any of your LIBC frontend. 

`fopencookie` is a solution/vilain patch which only regards end-user and a LIBC frontend to address the implementation of open_memstream and fmemopen, nuttx_file_struct should remain agnostic to such concerns, now I don't say it could be smarter and hold default callback pointers, the problem: the ones defined by `fopencookie` do not match exactly.

But again do not hot patch such.

xiaoxiang781216 commented:
> you would if you want to keep separated internal implementation to actual in-memory Os layout and a libc stdio frontend concern.
> 
> Indeed `nutxx/fs.h` is one possible core-stream handling implementation, you don't want to mixup ;

If you don't like file_struct is contained in nuttx/fs.h, the right fix is moving it to stdio.h, but don't create a new libc_stdio_file since libc_stdio_file lack many fields which is required to implement buffer I/O and fungetc etc.

> theoretically, you could swap to any `OS vfs` implementation without changing any of your LIBC frontend.
> 

`struct file_struct` is designed to work with stdio, `struct file` is designed to work with vfs. Please don't confuse them.

patacongo commented:
> @patacongo I think @michallenc initial design may confuse you. stdio is a pure userspace library, the implementation of fopencookie or fmemopen doesn't need involve any kernel stuff: it's enough to add the callback(read/write/seek/close) to struct file_struct(FILE).
> So, I would suggest you read the manual from:
> https://pubs.opengroup.org/onlinepubs/9699919799/functions/fmemopen.html
> https://man7.org/linux/man-pages/man3/fopencookie.3.html

AFAIK that is not true.  The user/supervisor mode is only one of several issues.  I have enumerated many others below that this design pure user-space design will not handle.  The bottom line:  You cannot call from one task/process address space into another.  In the FLAT build, that might work in some very, very simple cases.  But it will not work in general.  It will never work at all in the KERNEL build if the client and server lie in different process address spaces.  The virtual user address spaces overlap and it you try calling into it you will certainly crash.

In order to that successfully and reliably, a context switch is required.

I don't see any reason for me to participate in this any further.  I've tried to explain all of this as carefully as I can but I still see you guys marching ahead to fall off the cliff.

xiaoxiang781216 commented:
> > @patacongo I think @michallenc initial design may confuse you. stdio is a pure userspace library, the implementation of fopencookie or fmemopen doesn't need involve any kernel stuff: it's enough to add the callback(read/write/seek/close) to struct file_struct(FILE).
> > So, I would suggest you read the manual from:
> > https://pubs.opengroup.org/onlinepubs/9699919799/functions/fmemopen.html
> > https://man7.org/linux/man-pages/man3/fopencookie.3.html
> 
> AFAIK that is not true. The user/supervisor mode is only one of several issues. I have enumerated many others below that this design pure user-space design will not handle. The bottom line: You cannot call from one task/process address space into another. 

This case isn't suitable for stdio functions. How can you pass FILE * from one task/process to other task/process in kernel mode? It's impossible to happen in the first place.

patacongo commented:
> This case isn't suitable for stdio functions. How can you pass FILE * from one task/process to other task/process in kernel mode? So it's impossible to call the callback function cross the address space boundary since FILE* return from fopencookie can only be used by the caller's task/process and is always invalid and impossible to pass FILE* to another task/process at the first place.

I don't see the relevance of the FILE*.  I was more concerned at the io_funcs and cookie.  But looking at the prototype, it seems impossible that the io_funcs could like in a different address space. from the caller of fopencookie().

mu578 commented:
> > This case isn't suitable for stdio functions. How can you pass FILE * from one task/process to other task/process in kernel mode? So it's impossible to call the callback function cross the address space boundary since FILE* return from fopencookie can only be used by the caller's task/process and is always invalid and impossible to pass FILE* to another task/process at the first place.
> 
> I don't see the relevance of the FILE*. I was more concerned at the io_funcs and cookie. But looking at the prototype, it seems impossible that the io_funcs could like in a different address space. from the caller of fopencookie().

Aggreed, it would require to pound defined and only allowing such APIs in usermode. However, it could be doable by making a new vfs interface handling such but it would require a whole revamp and further thoughts to make it possible, as is : it is not hot patchable -> so my proposal to support indirection for user-mode build only, at least it can be undone and evolve without breaking.

That is impossible to callback from two distinct regions ; it would require some kind of trampoline jump with copies back and forth.

xiaoxiang781216 commented:
> > This case isn't suitable for stdio functions. How can you pass FILE * from one task/process to other task/process in kernel mode? So it's impossible to call the callback function cross the address space boundary since FILE* return from fopencookie can only be used by the caller's task/process and is always invalid and impossible to pass FILE* to another task/process at the first place.
> 
> I don't see the relevance of the FILE*. I was more concerned at the io_funcs and cookie. But looking at the prototype, it seems impossible that the io_funcs could like in a different address space. from the caller of fopencookie().

Yes, this is the key point: both io_funcs and cookie can only be used in the same proccess(address space), which simplify the thing a lot if the implementation can avoid the involve any kernel space stuff.

xiaoxiang781216 commented:
> > > This case isn't suitable for stdio functions. How can you pass FILE * from one task/process to other task/process in kernel mode? So it's impossible to call the callback function cross the address space boundary since FILE* return from fopencookie can only be used by the caller's task/process and is always invalid and impossible to pass FILE* to another task/process at the first place.
> > 
> > 
> > I don't see the relevance of the FILE*. I was more concerned at the io_funcs and cookie. But looking at the prototype, it seems impossible that the io_funcs could like in a different address space. from the caller of fopencookie().
> 
> Aggreed, it would require to pound defined and only allowing such APIs in usermode.

Of course, all FILE functions can only be called inside the usermode, NuttX kernel forbid kernel code to use any FILE functions.

> However, it could be doable by making a new vfs interface handling such but it would require a whole revamp and further thoughts to make it possible, as is : it is not hot patchable -> so my proposal to support indirection for user-mode build only, at least it can be undone and evolve without breaking.

@mu578 you may misunderstand NuttX file system:

- nuttx/fs/vfs is the kernel VFS infrastructure, which is part of kernel, it's hard and bad to call the callback provided by userspace
- nuttx/libc/stdio is userspace library, which is very simple and common to call the callback provided by userspace

> 
> That is impossible to callback from two distinct regions ; it would require some kind of trampoline jump with copies back and forth. E.g if `function pointers` were actually `just address pointers pointing to an actual somewhere` we would not have this conversation, but the reality is different.
> 

Do you want to implement the userspace file system? NuttX already support it by message queue:
https://github.com/apache/nuttx/blob/master/include/nuttx/fs/userfs.h

> Myself, I would go to support standard only: `open_memstream` and `fmemopen` ; there is no reason to expose a non standard `fopencookie` call ; the two requires some memory mapping that's it, no callback exposure. It is all about to be able to use FILE api on a user in-memory buffer instead of on a `file descriptor/ino` ; which in case of nuttx is `kif-kif bourricot` a mere detail.

If we want to support all open_memstream/open_wmemstream/fmemopen, it's important to setup a callback infrastructure otherwise fread/fwrite need add many:
```
if (file handle)
  {
    ...
  }
else if (mem handle)
  {
    ...
  }
else if (xxx handle)
  {
    ...
  }
```

michallenc commented:
To summarize the so far discussion a bit:

The problem discussed here is a that this implementation reuses `file_operation_s` infrastructure which is a part of a file system (thus kernel address space) which is my fault as I haven´t realize during the implementation that the reuse of current infrastructure would lead to this.

I quite like the proposition of reworked/new file system @mu578 suggested but as he mentioned, it is not a stuff for one patch and probably should be discussed on a mailing list first. I also do not agree with the statement 

> Myself, I would go to support standard only: open_memstream and fmemopen ; there is no reason to expose a non standard fopencookie call ; the two requires some memory mapping that's it, no callback exposure.

as `fopencookie` can be beneficial for programmers (maybe even more than `fmemopen`). And while not being POSIX, we should keep in mind the NuttX compatibility with Linux imho. Also if done correctly, `fopencookie` can be used both as libc call available for programmers and internal hook for `fopencookie` and `open_memstream`.

Regarding kernel protected builds, address space (and also priorities to keep strict scheduling), I think it should be sufficient to ensure user calls read/write/seek/close do not enter kernel space i.e. `fs/vfs` when used with `fopencookie`. In that case all code would remain in stdio libc library and would be purely user space therefore none of the mentioned issues should occur. I would go with @xiaoxiang781216 suggested implementation for that and we will see where it leads.

mu578 commented:
> > > > This case isn't suitable for stdio functions. How can you pass FILE * from one task/process to other task/process in kernel mode? So it's impossible to call the callback function cross the address space boundary since FILE* return from fopencookie can only be used by the caller's task/process and is always invalid and impossible to pass FILE* to another task/process at the first place.
> > > 
> > > 
> > > I don't see the relevance of the FILE*. I was more concerned at the io_funcs and cookie. But looking at the prototype, it seems impossible that the io_funcs could like in a different address space. from the caller of fopencookie().
> > 
> > 
> > Aggreed, it would require to pound defined and only allowing such APIs in usermode.
> 
> Of course, all FILE functions can only be called inside the usermode, NuttX kernel forbid kernel code to use any FILE functions.
> 
> > However, it could be doable by making a new vfs interface handling such but it would require a whole revamp and further thoughts to make it possible, as is : it is not hot patchable -> so my proposal to support indirection for user-mode build only, at least it can be undone and evolve without breaking.
> 
> @mu578 you may misunderstand NuttX file system:
> 
> * nuttx/fs/vfs is the kernel VFS infrastructure, which is part of kernel, it's hard and bad to call the callback provided by userspace
> * nuttx/libc/stdio is userspace library, which is very simple and common to call the callback provided by userspace
> 
> > That is impossible to callback from two distinct regions ; it would require some kind of trampoline jump with copies back and forth. E.g if `function pointers` were actually `just address pointers pointing to an actual somewhere` we would not have this conversation, but the reality is different.
> 
> Do you want to implement the userspace file system? NuttX already support it by socket: https://github.com/apache/nuttx/blob/master/include/nuttx/fs/userfs.h
> 
> > Myself, I would go to support standard only: `open_memstream` and `fmemopen` ; there is no reason to expose a non standard `fopencookie` call ; the two requires some memory mapping that's it, no callback exposure. It is all about to be able to use FILE api on a user in-memory buffer instead of on a `file descriptor/ino` ; which in case of nuttx is `kif-kif bourricot` a mere detail.
> 
> If we want to support all open_memstream/open_wmemstream/fmemopen, it's important to setup a callback infrastructure otherwise fread/fwrite need to add many if/else:
> 
> ```
> if (file handle)
>   {
>     ...
>   }
> else if (mem handle)
>   {
>     ...
>   }
> else if (xxx handle)
>   {
>     ...
>   }
> ```
> 
> If we need to add the callback to FILE, why not follow fopencookie which exist in most libc implementation(glibc, newlib, *BSD), but add a NuttX specific function instead? I don't see any benefit.

I fully understand vfs.

xiaoxiang781216 commented:
> I fully understand vfs, no callback infrastructure is needed,

> that's a glibc patch, who cares?

But I care where the implementation is elegance and extendable.  The code like spaghetti should avoid as mention in:
https://github.com/apache/nuttx/blob/master/INVIOLABLES.md?plain=1#L36-L41
https://github.com/apache/nuttx/blob/master/INVIOLABLES.md?plain=1#L90-L94

> just deal with a flag about memory provider that's it.

even for memory stream, there are at least two cases:

1. fmemopen work on a fix size array(provided by caller, or allocated by self)
2. open_memstream work on an allocate/reallocate extendable memory

The detail is evil, you have to add more flags/fields and many if/else in many common places to handle the different behavior required by spec. The callback is better design here to encapsulate the difference behind the differerent callback(fd vs. mem, fixed vs. dynamically).

mu578 commented:
> * provided by caller, or allocated by self)
> * open_memstream work on an allocate/reallocate extendable memor

yes make a context buffer |internal|user_malloc|user_dont_touch|

```

_io_buffer_grow_if_needed(stream_flag, ** buffer)
{
  if this type -> do that or not do
}

_io_buffer_free_if_needed(stream_flag, ** buffer)
{
  if this type -> do that or not do
}

```

michallenc commented:
I have updated the implementation with internal callbacks so now the code should entirely avoid kernel address space. How are we going to design some tests that would ensure the correctness?

michallenc commented:
Some initial tests can be found here: https://github.com/michallenc/incubator-nuttx-apps/commit/1cd88d54c7119fad000866a3a62f8c5afad2ebc8 I have only tested it in FLAT mode so far as I have not been able to compile NuttX in protected mode (regardless of `fopencookie`, I tried upstream NuttX without changes as well) on both samv7 and imxrt MCUs. SAMv7 fails on `NVIC_CFAULTS_MUNSTKERR` (MemManage fault on unstacking) and I could not get any output from imxrt. I used @acassis video on https://www.youtube.com/watch?v=p2aXmPNUnCg for configuration, so that stepts should be alright. There might be something wrong with MPU for ARMv7-M...

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -425,6 +425,55 @@ struct inode
+ * off_t is int64_t if CONFIG_FS_LARGEFILE is defined and int32_t if not.
+ */
+
+typedef ssize_t cookie_read_function_t(void *cookie, char *buf, size_t size);
```
add FAR to ALL pointers

michallenc replied:
Done

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -425,6 +425,55 @@ struct inode
+typedef ssize_t cookie_read_function_t(void *cookie, char *buf, size_t size);
+typedef ssize_t cookie_write_function_t(void *cookie, const char *buf,
+                                        size_t size);
+typedef off_t cookie_seek_function_t(void *cookie, off_t *offset,
```
```suggestion
typedef CODE int cookie_seek_function_t(FAR void *cookie, FAR off_t *offset,
```
add CODE to ALL function typdef

michallenc replied:
done

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -425,6 +425,55 @@ struct inode
+                                     int whence);
+typedef int cookie_close_function_t(void *cookie);
+
+typedef struct
```
```suggestion
typedef struct cookie_io_functions_s
```

michallenc replied:
done

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -425,6 +425,55 @@ struct inode
+  FAR write_internal_cb_t *write;
+  FAR seek_internal_cb_t *seek;
+  FAR close_internal_cb_t *close;
+} io_callbacks_t;
```
why define the duplicated typedef and struct(from line 463 to line 475), but use cookie_io_functions_t directly?

michallenc replied:
Removed

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -497,6 +546,8 @@ struct file_struct
   rmutex_t                fs_lock;      /* Recursive lock */
   int                     fs_fd;        /* File descriptor associated with stream */
+  FAR struct cookie_s     fs_cookie;    /* Cookie structure for fopencookie */
+  io_callbacks_t          fs_callbacks; /* Internal callbacks for IO call */
```
duplicated with fs_cookie.cookie_io

michallenc replied:
I would not put internal callbacks to the same structure as fopencookie callbacks simply for code clarity. Also we still need those structures twice... one for internal callback and other for user defined callback.

michallenc replied:
Removed

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -497,6 +546,8 @@ struct file_struct
   FAR struct file_struct *fs_next;      /* Pointer to next file stream */
   rmutex_t                fs_lock;      /* Recursive lock */
   int                     fs_fd;        /* File descriptor associated with stream */
```
fs_fd should union with fs_cookie.cookie

michallenc replied:
This should not be needed since I would pass the entire stream to internal callback.

xiaoxiang781216 replied:
but the current design waste the memory:

1. fs_fd dup with fs_cookie.cookie
2. fs_callbacks dup with fs_cookie.cookie_io

total bytes equal to 20(32bit) or 36(64bit).

michallenc replied:
I removed  `fs_fd` entirely as it is not needed anymore. Instead all file descriptors are passed through `fs_cookie` (which is used for user cookie if fopencookie is used). `fs_callbacks` removed.

xiaoxiang781216 commented on libs/libc/libc.h:
```diff
@@ -245,6 +245,13 @@ bool lib_isbasedigit(int ch, int base, FAR int *value);
 
+/* Defined in lib_stdio_cb */
+
+ssize_t lib_fread_cb(void *stream, char *buf, size_t size);
```
add FAR for ALL pointer

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/libc.h:
```diff
@@ -245,6 +245,13 @@ bool lib_isbasedigit(int ch, int base, FAR int *value);
+
+ssize_t lib_fread_cb(void *stream, char *buf, size_t size);
+ssize_t lib_fwrite_cb(void *stream, const char *buf, size_t size);
+off_t lib_fseek_cb(void *stream, off_t offset, int whence);
```
```suggestion
int lib_fseek_cb(void *stream, off_t offset, int whence);
```

michallenc replied:
fseek for fopencookie is defined with off_t

xiaoxiang781216 replied:
Yes, but fseek could do the conversion, so we don't need seek_internal_cb_t.

michallenc replied:
Yes, but `lib_fseek_cb` passed to `cookie_seek_function_t` which has to be defined with `off_t` to correspond with fopencookie specification.

xiaoxiang781216 commented on libs/libc/libc.h:
```diff
@@ -245,6 +245,13 @@ bool lib_isbasedigit(int ch, int base, FAR int *value);
 
 int lib_checkbase(int base, FAR const char **pptr);
 
+/* Defined in lib_stdio_cb */
```
```suggestion
/* Defined in lib_stdio_cb.c */
```

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_fclose.c:
```diff
@@ -129,7 +129,7 @@ int fclose(FAR FILE *stream)
                                                (uintptr_t)stream);
           status = android_fdsan_close_with_tag(stream->fs_fd, tag);
```
need move fdscan related code to stdio callback

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_fclose.c:
```diff
@@ -129,7 +129,7 @@ int fclose(FAR FILE *stream)
           status = android_fdsan_close_with_tag(stream->fs_fd, tag);
 #else
-          status = close(stream->fs_fd);
+          status = stream->fs_callbacks.close(stream);
```
need pass cookie instead stream here

michallenc replied:
as mentioned above, I would pass stream

michallenc replied:
Done as suggested.

xiaoxiang781216 commented on libs/libc/stdio/lib_stdio_cb.c:
```diff
@@ -0,0 +1,76 @@
+ssize_t lib_fwrite_cb(FAR void *s, const char *buf, size_t size)
+{
+  FAR FILE *stream = (FAR FILE *)s;
+  return _NX_WRITE(stream->fs_fd, buf, size);
```
change _NX_WRITE to write

michallenc replied:
_NX_WRITE is either write or nx_write based on which mode we are in and whether this is kernel space or user space. STDIO for protected build should be compiled twice, once for kernel and once for user space, therefore macro should decide which function is called.

xiaoxiang781216 replied:
stdio shouldn't be called inside kernel space, so _NX_WRITE isn't needed.

michallenc replied:
Sorry for the delay @xiaoxiang781216, I was on a vacation without notebook.

NuttX documentation states

> Normally the logic in this file builds to a single library (libc.a). However, if NuttX is built as a separately compiled kernel (with CONFIG_BUILD_PROTECTED=y or CONFIG_BUILD_KERNEL=y), then the contents of this directory are built as two libraries: One for use by user programs (libc.a) and one for use only within the <kernel> space (libkc.a).

Therefore stdio can be called inside kernel space and `_NX_WRITE` should be there to decide.

xiaoxiang781216 replied:
Ok, let's keep it.

xiaoxiang781216 commented on libs/libc/stdio/lib_stdio_cb.c:
```diff
@@ -0,0 +1,76 @@
+ssize_t lib_fread_cb(FAR void *s, char *buf, size_t size)
+{
+  FAR FILE *stream = (FAR FILE *)s;
+  return _NX_READ(stream->fs_fd, buf, size);
```
change _NX_READ to read

michallenc replied:
As above

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,181 @@
+  return ret;
+}
+
+static int fclosecookie(FAR void *s)
```
let's remove f*cookie and forward to cookie_io_functions_t directly

michallenc replied:
Done

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,181 @@
+{
+  FAR FILE *stream;
+
+  stream = lib_zalloc(sizeof(FILE));
```
stream need allocate through fs_fdopen like fopen/fdopen too.

michallenc replied:
I cannot call fs_fdopen without valid fd.

xiaoxiang781216 replied:
In the new design, FILE doesn't couple with fd, so fs_fdopen should change the first argument from fd to cookie.

michallenc replied:
Done.

michallenc commented:
@xiaoxiang781216 I have updated the implementation and implemented most of your suggestions (see separate comments). There is a new PR in nuttx-apps [add fopecookie test and use fileno interface instead of accessing FILE](https://github.com/apache/nuttx-apps/pull/2112) that adds test application and also fix some `FILE` usage in `system` (this is should be merged before NuttX patch because of CI).

xiaoxiang781216 commented on sched/group/group_setupstreams.c:
```diff
@@ -64,9 +64,9 @@ int group_setupstreams(FAR struct task_tcb_s *tcb)
-  fs_fdopen(1, O_WROK | O_CREAT, (FAR struct tcb_s *)tcb, NULL);
-  fs_fdopen(2, O_WROK | O_CREAT, (FAR struct tcb_s *)tcb, NULL);
+  fs_fdopen(0, 0, O_RDONLY,         (FAR struct tcb_s *)tcb, NULL);
+  fs_fdopen(1, 0, O_WROK | O_CREAT, (FAR struct tcb_s *)tcb, NULL);
```
```suggestion
  fs_fdopen(1, 1, O_WROK | O_CREAT, (FAR struct tcb_s *)tcb, NULL);
```

xiaoxiang781216 commented on sched/group/group_setupstreams.c:
```diff
@@ -64,9 +64,9 @@ int group_setupstreams(FAR struct task_tcb_s *tcb)
-  fs_fdopen(2, O_WROK | O_CREAT, (FAR struct tcb_s *)tcb, NULL);
+  fs_fdopen(0, 0, O_RDONLY,         (FAR struct tcb_s *)tcb, NULL);
+  fs_fdopen(1, 0, O_WROK | O_CREAT, (FAR struct tcb_s *)tcb, NULL);
+  fs_fdopen(2, 0, O_WROK | O_CREAT, (FAR struct tcb_s *)tcb, NULL);
```
```suggestion
  fs_fdopen(2, 2, O_WROK | O_CREAT, (FAR struct tcb_s *)tcb, NULL);
```

xiaoxiang781216 commented on libs/libc/stdio/lib_stdio_cb.c:
```diff
@@ -0,0 +1,90 @@
+
+#ifdef CONFIG_FDSAN
+  uint64_t tag;
+  tag = android_fdsan_create_owner_tag(ANDROID_FDSAN_OWNER_TYPE_FILE,
```
```suggestion
  uint64_t tag = android_fdsan_create_owner_tag(ANDROID_FDSAN_OWNER_TYPE_FILE,
```

michallenc replied:
Done

xiaoxiang781216 commented on libs/libc/stdio/lib_stdio_cb.c:
```diff
@@ -0,0 +1,90 @@
+
+int lib_fclose_cb(FAR void *cookie)
+{
+  int fd = (int)(intptr_t)cookie;
```
why check cookie here

michallenc replied:
There should be a check whether we have valid fd before we call `close()`. This was previously in `lib_fclose.c` but I moved it here since in `lib_fclose.c` we dont know whether it is fd or cookie.

xiaoxiang781216 replied:
close will do the check for us

xiaoxiang781216 commented on libs/libc/stdio/lib_libstream.c:
```diff
@@ -66,13 +67,27 @@ void lib_stream_initialize(FAR struct task_group_s *group)
   /* Initialize stdin, stdout and stderr stream */
 
-  list->sl_std[0].fs_fd = -1;
+  list->sl_std[0].fs_cookie       = (FAR void *)(intptr_t)fd;
```
-1 and remove line 59&70

michallenc replied:
Done

xiaoxiang781216 commented on libs/libc/stdio/lib_fopen.c:
```diff
@@ -120,7 +127,7 @@ FAR FILE *fopen(FAR const char *path, FAR const char *mode)
   if (fd >= 0)
     {
-      ret = fs_fdopen(fd, oflags, NULL, &filep);
+      ret = fs_fdopen(fd, 0, oflags, NULL, &filep);
```
```suggestion
      ret = fs_fdopen(fd, fd, oflags, NULL, &filep);
```

xiaoxiang781216 commented on libs/libc/stdio/lib_fopen.c:
```diff
@@ -74,7 +74,7 @@ FAR FILE *fdopen(int fd, FAR const char *mode)
   if (oflags >= 0)
     {
-      ret = fs_fdopen(fd, oflags, NULL, &filep);
+      ret = fs_fdopen(fd, 0, oflags, NULL, &filep);
```
```suggestion
      ret = fs_fdopen(fd, fd, oflags, NULL, &filep);
```

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,83 @@
+
+  /* Call fs_fdopen with invalid file descriptor and cookie flag set to 1 */
+
+  ret = fs_fdopen(-1, 1, oflags, NULL, &filep);
```
```suggestion
  ret = fs_fdopen(-1, cookie, oflags, NULL, &filep);
```

xiaoxiang781216 commented on libs/libc/stdio/lib_libfread_unlocked.c:
```diff
@@ -170,7 +170,9 @@ ssize_t lib_fread_unlocked(FAR void *ptr, size_t count, FAR FILE *stream)
                     {
-                      bytes_read = _NX_READ(stream->fs_fd, dest, remaining);
+                      bytes_read = stream->fs_iofunc.read(stream->fs_cookie,
+                                                           (char *)dest,
```
```suggestion
                                                          (char *)dest,
```

michallenc replied:
Done

xiaoxiang781216 commented on libs/libc/stdio/lib_libfread_unlocked.c:
```diff
@@ -213,9 +215,9 @@ ssize_t lib_fread_unlocked(FAR void *ptr, size_t count, FAR FILE *stream)
-                                            stream->fs_bufread,
-                                            buffer_available);
+                      bytes_read = stream->fs_iofunc.read(stream->fs_cookie,
+                                              (char *)stream->fs_bufread,
```
```suggestion
                                                             (char *)stream->fs_bufread,
```

michallenc replied:
I can´t move it there, the line would be too long.

xiaoxiang781216 replied:
ok

xiaoxiang781216 replied:
but, need add FAR

xiaoxiang781216 replied:
could we change fs_bufread type to char * to avoid the cast

michallenc replied:
FAR added. I think I would rather cast it here than change the type through the entire NuttX now. Maybe in some further patch but not together with fopencookie.

michallenc commented on fs/vfs/fs_fdopen.c:
```diff
@@ -120,7 +120,7 @@ int fs_fdopen(int fd, int oflags, FAR struct tcb_s *tcb,
   /* Check input parameters */
 
-  if (fd < 0)
+  if (fd < 0 && cookie == 0)
```
@xiaoxiang781216 This is basically just a information passed to fs_fdopen that initialization is done for fopencookie (1 true) and not standard stream (0 false). It is not to pass the pointer to cookie. The implementation with pointer to cookie would be tricky because we need valid fd to pass the checks (and we need -1 reserved if open call returns error). This adds two more if statements but does not complicate the code more. Applies to most other comments.

xiaoxiang781216 replied:
with https://github.com/apache/nuttx/pull/10913, you could move FILE allocation into fopencookie, and forward fdopen to fopencookie.

michallenc commented on include/nuttx/fs/fs.h:
```diff
@@ -425,6 +425,31 @@ struct inode
+ * or to standard file system functions if not.
+ */
+
+typedef CODE ssize_t cookie_read_function_t(void *cookie, char *buf,
```
Done

xiaoxiang781216 replied:
but, not change

michallenc replied:
Done, seems i havent commited it.

xiaoxiang781216 commented:
> @xiaoxiang781216 I have updated the implementation and implemented most of your suggestions (see separate comments). There is a new PR in nuttx-apps [add fopecookie test and use fileno interface instead of accessing FILE](https://github.com/apache/nuttx-apps/pull/2112) that adds test application and also fix some `FILE` usage in `system` (this is should be merged before NuttX patch because of CI).

@michallenc but, the ci still fail, could you take a look?

michallenc commented:
> @michallenc but, the ci still fail, could you take a look?

Yes, I fixed some on Saturday but I suppose the rest is somehow related to change to `fs_fdopen` which should be (at least partially) fixed by the update from your patch a remove of `fs_fdopen`.

michallenc commented on libs/libc/stdio/lib_fclose.c:
```diff
@@ -115,32 +115,41 @@ int fclose(FAR FILE *stream)
-          /* Close the file descriptor and save the return status */
+          /* This is a little warkaround. Ideally we would want to call
+           * fs_iofunc.close and let that handle it. FDSAN however requires
+           * stream to create owner tag and this is not avalialble in
```
Done

xiaoxiang781216 commented:
> > @michallenc but, the ci still fail, could you take a look?
> 
> Yes, I fixed some on Saturday but I suppose the rest is somehow related to change to `fs_fdopen` which should be (at least partially) fixed by the update from your patch a remove of `fs_fdopen`. I will update the patch once yours is merged.

@michallenc https://github.com/apache/nuttx/pull/10913 was merged, you can rebase your patch on top of it now.

michallenc commented:
> @michallenc #10913 was merged, you can rebase your patch on top of it now.

@xiaoxiang781216 Updated. Local CI runs were successful, lets see what GitHub CI does.

xiaoxiang781216 commented on libs/libc/stdio/lib_fclose.c:
```diff
@@ -115,32 +115,18 @@ int fclose(FAR FILE *stream)
-#else
-          status = close(stream->fs_fd);
-#endif
+      status = stream->fs_iofunc.close(stream->fs_cookie);
```
should we check close != NULL before calling, or fill the empty callback function in fopencookie

michallenc replied:
Added NULL check. Fopencookie actually allows the user to have close == NULL.

xiaoxiang781216 commented on include/nuttx/fs/fs.h:
```diff
@@ -426,6 +426,31 @@ struct inode
+typedef CODE ssize_t cookie_write_function_t(void *cookie, const char *buf,
+                                             size_t size);
+typedef CODE off_t cookie_seek_function_t(void *cookie, off_t *offset,
+                                     int whence);
```
align

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_libfflush.c:
```diff
@@ -107,7 +107,9 @@ ssize_t lib_fflush_unlocked(FAR FILE *stream, bool bforce)
 
-          bytes_written = _NX_WRITE(stream->fs_fd, src, nbuffer);
+          bytes_written = stream->fs_iofunc.write(stream->fs_cookie,
+                                                  (const char *)src,
```
can we change the type at line 62 to avoid the cast

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_libfread_unlocked.c:
```diff
@@ -170,7 +170,9 @@ ssize_t lib_fread_unlocked(FAR void *ptr, size_t count, FAR FILE *stream)
                     {
-                      bytes_read = _NX_READ(stream->fs_fd, dest, remaining);
+                      bytes_read = stream->fs_iofunc.read(stream->fs_cookie,
+                                                          (char *)dest,
```
can we change the type at line 49 to avoid the cast

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_setvbuf.c:
```diff
@@ -133,7 +134,7 @@ int setvbuf(FAR FILE *stream, FAR char *buffer, int mode, size_t size)
   /* Return EBADF if the file is not open */
 
-  if (stream->fs_fd < 0)
+  if (fd < 0)
```
should we allow for cookie case

michallenc replied:
hmm, what about this? Can we presume that it is not possible to have stream structure without the stream actually being open?

xiaoxiang781216 replied:
It's impossible to handle this type of error reliably. It's caller's responsibility to pass stream which comes from fopen/fdopen/fopencookie.

michallenc replied:
Ok, removed.

xiaoxiang781216 commented on libs/libc/stdio/lib_ungetc.c:
```diff
@@ -52,7 +53,7 @@ int ungetc(int c, FAR FILE *stream)
   /* Stream must be open for read access */
 
-  if ((stream->fs_fd < 0) || ((stream->fs_oflags & O_RDOK) == 0))
+  if ((fd < 0) || ((stream->fs_oflags & O_RDOK) == 0))
```
should we allow cookie case? which may pass in the negative value

michallenc replied:
done

michallenc commented:
Hmm, sim-01 fails without actually logging any error.

xiaoxiang781216 commented on sched/tls/task_initinfo.c:
```diff
@@ -28,6 +28,8 @@
 #include <nuttx/kmalloc.h>
 #include <nuttx/mutex.h>
 
+#include <nuttx/lib/lib.h>
```
change to libc.h

xiaoxiang781216 replied:
Yes, forget it's in side kernel space, but kernel code can't reference libc function directly even th inclusion is added. The better fix should be that seek/fwrite/fread call lseek/_NX_WRITE/_NX_READ directly if the related callback is NULL.

xiaoxiang781216 replied:
@michallenc this is the last comment you need consider, otherwise the change will break protected and kernel mode.

xiaoxiang781216 replied:
> `<nuttx/lib/lib.h>` was used to reference `lib_stream_initialize` and `lib_stream_release` before your patch so I am not sure if it is such a problem in this case.

both functions are moved into sched, so the inclusion doesn't need anymore

> Deciding what function is to be called would require some if statements in running code which I would like to avoid.

but you have to do th check since:

1. stdin/stdout/sderr is initialized inside kernel space
2. The assignment of stdin/stdout/stderr's callback inside kernel space point to the lib_fxxx_cb inside kernel space
3. fopen/fdopen point lib_fxxx_cb in user space

when user call fwrite/fread on stdin/stdout/stderr, prefetch abort will happen since user space doesn't right to execute the code inside the kernel space.

xiaoxiang781216 commented on libs/libc/stdio/lib_libfwrite.c:
```diff
@@ -132,7 +132,8 @@ ssize_t lib_fwrite_unlocked(FAR const void *ptr, size_t count,
   if (count >= CONFIG_STDIO_BUFFER_SIZE)
     {
-      ret = _NX_WRITE(stream->fs_fd, src, count);
+      ret = stream->fs_iofunc.write(stream->fs_cookie, (const char *)src,
```
change src type to avoid the cast

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_libfread_unlocked.c:
```diff
@@ -258,7 +260,9 @@ ssize_t lib_fread_unlocked(FAR void *ptr, size_t count, FAR FILE *stream)
             {
-              bytes_read = _NX_READ(stream->fs_fd, dest, remaining);
+              bytes_read = stream->fs_iofunc.read(stream->fs_cookie,
+                                                  (char *)dest,
```
ditto

xiaoxiang781216 commented on libs/libc/stdio/lib_libfgets.c:
```diff
@@ -97,10 +97,11 @@ FAR char *lib_fgets_unlocked(FAR char *buf, size_t buflen, FILE *stream,
   /* Sanity checks */
 
-  if (!stream || !buf || stream->fs_fd < 0)
+  if (!stream || !buf || fd < 0)
```
should we allow cookie < 0?

xiaoxiang781216 replied:
Yes, it's better to let the code which really use fd to do the check.

xiaoxiang781216 commented:
please fix:
```
81a82
> CONFIG_TESTING_FOPENCOOKIE_TEST=y
87d87
< CONFIG_TESTING_FOPENCOOKIE_TEST=y
Saving the new configuration file
HEAD detached at pull/10602/merge
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   boards/sim/sim/sim/configs/citest/defconfig
```

xiaoxiang781216 commented on libs/libc/stdio/lib_fclose.c:
```diff
@@ -64,6 +64,7 @@ int fclose(FAR FILE *stream)
   int errcode = EINVAL;
   int ret = ERROR;
   int status;
+  int fd = (int)(intptr_t)stream->fs_cookie;
```
move after line 126 since stream may be NULL.

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,153 @@
+ * Name: cookie_read_cb
+ ****************************************************************************/
+
+ssize_t cookie_read_cb(FAR void *cookie, FAR char *buf, size_t size)
```
add static for ALL dummy function.

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,153 @@
+{
+  /* Per specification: if *read is a null pointer, then reads from the
+   * custom stream always return end of file.
+   */
```
so, we should return 0 without touching errno

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,153 @@
+ssize_t cookie_write_cb(FAR void *cookie, FAR const char *buf, size_t size)
+{
+  /* Per specification: if *write is a null pointer, then output to the
+   * stream is discarded.
```
so, we should return size

michallenc replied:
good spot, done

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,153 @@
+   * possible to perform seek operations on the stream.
+   */
+
+  errno = -EINVAL;
```
-ENOSUP or -ENOSYS?

michallenc replied:
-ENOTSUP

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,153 @@
+{
+  return OK;
+}
+
```
add public section banner

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_ungetc.c:
```diff
@@ -52,7 +52,7 @@ int ungetc(int c, FAR FILE *stream)
   /* Stream must be open for read access */
 
-  if ((stream->fs_fd < 0) || ((stream->fs_oflags & O_RDOK) == 0))
+  if (((stream->fs_oflags & O_RDOK) == 0))
```
remove the extra ()

michallenc replied:
done

xiaoxiang781216 commented on libs/libc/stdio/lib_libfread_unlocked.c:
```diff
@@ -46,13 +46,14 @@
   size_t gulp_size;
   int ret;
 #endif
+  int fd = (int)(intptr_t)stream->fs_cookie;
```
need move to line 72 since we can known stream 
isn't NULL only after line 60

michallenc replied:
Done

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,156 @@
+   * stream is discarded.
+   */
+
+  return 0;
```
need return size here

michallenc replied:
We should return no bytes written, no?

xiaoxiang781216 replied:
0 mean nothing is written, but the spec requires discard all input from the comment.

michallenc replied:
Hm, by discarding all input I understand that nothing is written. Returning size would mean the application could expect the pointer is somewhere else then it really is. I will look at other libc implementations how they deal with this.

michallenc replied:
Ok, musl also returns size. Changed.

xiaoxiang781216 commented on libs/libc/stdio/lib_libfread_unlocked.c:
```diff
@@ -46,13 +46,14 @@
 ssize_t lib_fread_unlocked(FAR void *ptr, size_t count, FAR FILE *stream)
 {
-  FAR unsigned char *dest = (FAR unsigned char *)ptr;
+  FAR char *dest = (FAR char *)ptr;
```
remove the cast?

michallenc replied:
Yes

xiaoxiang781216 commented on libs/libc/stdio/lib_fopencookie.c:
```diff
@@ -0,0 +1,156 @@
+   * possible to perform seek operations on the stream.
+   */
+
+  errno = -ENOTSUP;
```
set_errno(-ENOTSUP)?
